### PR TITLE
ref(pkg/*): move envoy proxy image to OSM's ConfigMap

### LINKED ISSUE #3135: Move envoy proxy sidecar image option to OSM's ConfigMap/CRD
**Please describe the Improvement and/or Feature Request**

Currently `osm-injector` accepts information regarding the sidecar as a command line argument:
https://github.com/openservicemesh/osm/blob/9ea49cfb762ed6433960e05719974040f5534660/charts/osm/templates/osm-injector-deployment.yaml#L37-L38

This issue proposes to move this into OSM's ConfigMap/CRD so that if the image needs to be patched for whatever reason, it can be done without patching the `osm-injector` deployment.

The image should be made available via the `Configurator` interface in `pkg/configurator` and used to set the envoy container image.
Update
https://github.com/openservicemesh/osm/blob/9ea49cfb762ed6433960e05719974040f5534660/pkg/injector/envoy_container.go#L37
to
`Image:           cfg.GetEnvoyImage()` 

Parent issue: https://github.com/openservicemesh/osm/issues/2801

**Scope (please mark with X where applicable)**
- New Functionality          [ ]
- Install                    [X]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [ ]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- CI System                  [ ]
- Demo                       [ ]
- Project Release            [ ]

### PR DESCRIPTION
**Pull request by jaellio**
**Description**:

Moves the envoy proxy image information from the osm-injector
deployment to OSM's ConfigMap. This change allows the image to be
patched without having to patch the osm-injector deployment. 
Resolves #3135.

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ X ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Demo                   [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
no

### CODE REVIEW

**Review by shashankram — APPROVED** [review]
Looks good. You might want to update `docs/content/docs/osm_config_map.md`.

**Review by snehachhabria — APPROVED** [review]

File: charts/osm/templates/osm-configmap.yaml [hunk 1]
```diff
@@ -9,6 +9,7 @@ data:
   permissive_traffic_policy_mode: {{ .Values.OpenServiceMesh.enablePermissiveTrafficPolicy | default "false" | quote }}
   egress: {{ .Values.OpenServiceMesh.enableEgress | quote }}
   envoy_log_level: {{ .Values.OpenServiceMesh.envoyLogLevel | quote }}
+  envoy_image: {{ .Values.OpenServiceMesh.sidecarImage | quote }}
```
Comment (allenlsy) [inline charts/osm/templates/osm-configmap.yaml:12]: I think everything adding to osm-configmap.yaml now needs to be sync'ed in MeshConfig CRD and struct definition. 

cc. @shashankram

**Review by eduser25 — APPROVED** [review]

File: charts/osm/templates/osm-configmap.yaml (hunk 1 above)
Comment (shashankram) [reply charts/osm/templates/osm-configmap.yaml:12]: @jaellio Would you mind adding this to the [MeshConfig CRD](https://github.com/openservicemesh/osm/blob/main/charts/osm/crds/meshconfig.yaml) and [client code](https://github.com/openservicemesh/osm/blob/main/pkg/apis/config/v1alpha1/mesh_config.go) in a subsequent change. The autogen client code needs to be regenerated using `./codegen/gen-crd-client.sh`.

File: pkg/apis/config/v1alpha1/mesh_config.go [hunk 2]
```diff
@@ -25,6 +25,7 @@ type MeshConfigSpec struct {
 type SidecarSpec struct {
 	EnablePrivilegedInitContainer bool   `json:"enablePrivilegedInitContainer,omitempty" yaml:"enablePrivilegedInitContainer,omitempty"`
 	LogLevel                      string `json:"logLevel,omitempty" yaml:"logLevel,omitempty"`
+	EnvoyImage                    string `json:"envoyImage,omitempty" yaml:"logLevel,omitempty"`
```
Comment (allenlsy) [inline pkg/apis/config/v1alpha1/mesh_config.go:28]: nit: `yaml:"envoyImage,omitempty"`

File: charts/osm/templates/osm-configmap.yaml (hunk 1 above)
Comment (jaellio) [reply charts/osm/templates/osm-configmap.yaml:12]: The changes have been added to this PR

File: cmd/cli/install.go [hunk 3]
```diff
@@ -347,6 +352,11 @@ func (i *installCmd) validateOptions() error {
 		return err
 	}
 
+	// validate the envoy log level type
```
Comment (snehachhabria) [inline cmd/cli/install.go:355]:
```suggestion
	// validate the envoy image
```

File: cmd/cli/install_test.go [hunk 4]
```diff
@@ -527,6 +528,29 @@ var _ = Describe("Test envoy log level types", func() {
...
+			Expect(err).NotTo(HaveOccurred())
+
+			err = isValidEnvoyImage("envoyproxy/envoy-alpine:v1.17.1")
+			Expect(err).NotTo(HaveOccurred())
+
+			err = isValidEnvoyImage("envoyproxy/envoy-alpine:v1.17.2")
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Should correctly error for invalid envoy log level", func() {
```
Comment (snehachhabria) [inline cmd/cli/install_test.go:544]:
```suggestion
		It("Should correctly error for invalid form of envoy image ", func() {
```

**Review by snehachhabria — COMMENTED** [review]
You will need to add a default value for the envoy image here as well : https://github.com/openservicemesh/osm/blob/663e1dafe0761e47ce7500d9dd7e98ab0557f757/cmd/init-osm-controller/init-osm-controller.go#L56

File: cmd/cli/install.go [hunk 5]
```diff
@@ -190,6 +193,7 @@ func newInstallCmd(config *helm.Configuration, out io.Writer) *cobra.Command {
 	f.BoolVar(&inst.deployJaeger, "deploy-jaeger", defaultDeployJaeger, "Deploy Jaeger in the namespace of the OSM controller")
 	f.StringVar(&inst.envoyLogLevel, "envoy-log-level", defaultEnvoyLogLevel,
 		fmt.Sprintf("Envoy log level is the level of logs collected from envoy and needs to be one of: %s", strings.Join(possibleEnvoyLogLevels, ",")))
+	f.StringVar(&inst.envoyImage, "envoy-image", defaultEnvoyImage, "Envoy image is the Envoy proxy sidecare image")
```
Comment (shashankram) [inline cmd/cli/install.go:196]: We should not be adding new CLI options here going forward, we have a pending issue to use `--set`: https://github.com/openservicemesh/osm/issues/3124

**Review by shashankram — CHANGES_REQUESTED** [review]
Hey @jaellio, could you remove the `envoy-image` option from the install CLI? This should not be necessary because chart values can be set during install using the `--set` option.

File: cmd/cli/install_test.go (hunk 4 above)
Comment (jaellio) [reply cmd/cli/install_test.go]: Removed based on the comment from @shashankram to remove the `envoy-image` option from the install CLI

File: docs/content/docs/osm_config_map.md [hunk 6]
```diff
@@ -93,6 +95,7 @@ use_https_ingress: must be a boolean
 |-----|------|---------------|--------------------------------|
 | enable_debug_server | bool | `"true"` | `kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"enable_debug_server":"false"}}' --type=merge` |
 | envoy_log_level | string | `"error"` | `kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"envoy_log_level":"info"}}' --type=merge` |
+| envoy_image | string | `"envoyproxy/envoy-alpine:v1.17.2"` | `kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"envoy_image":"envoyproxy/envoy-alpine:v1.17.1"}}' --type=merge` |
```
Comment (shashankram) [inline docs/content/docs/osm_config_map.md:98]:
```suggestion
| envoy_image | string | `"envoyproxy/envoy-alpine:v1.17.2"` | `kubectl patch ConfigMap osm-config -n osm-system -p '{"data":{"envoy_image":"envoyproxy/envoy-alpine:v1.17.2"}}' --type=merge` |
```

File: docs/content/docs/osm_config_map.md [hunk 7]
```diff
@@ -119,6 +122,7 @@ kubectl describe validatingwebhookconfiguration osm-webhook-osm
 | enable_debug_server | `must be a boolean` |
 | enable_privileged_init_container| `must be a boolean` |
 | envoy_log_level | `invalid log level` |
+| envoy_image | `must be of the form envoyproxy/envoy-alpine:vx.xx.x`
```
Comment (shashankram) [inline docs/content/docs/osm_config_map.md:125]:
```suggestion
| envoy_image | `must be of the form envoyproxy/envoy-alpine:v<major>.<minor>.<patch>`
```

File: pkg/configurator/validating_webhook.go [hunk 8]
```diff
@@ -55,6 +56,9 @@ const (
 	// mustBeValidLogLvl is the reason for denial for envoy_log_level field
 	mustBeValidLogLvl = ": invalid log level"
 
+	// mustBeValidLogLvl is the reason for denial for envoy_image field
+	mustBeValidEnvoyImage = ": must be of the form envoyproxy/envoy-alpine:vx.xx.x"
```
Comment (shashankram) [inline pkg/configurator/validating_webhook.go:60]:
```suggestion
	mustBeValidEnvoyImage = ": must be of the form envoyproxy/envoy-alpine:v<major>.<minor>.<patch>"
```

**Review by shashankram — APPROVED** [review]
Looks good, thanks!

File: pkg/apis/config/v1alpha1/mesh_config.go [hunk 9]
```diff
@@ -25,6 +25,7 @@ type MeshConfigSpec struct {
 type SidecarSpec struct {
 	EnablePrivilegedInitContainer bool   `json:"enablePrivilegedInitContainer,omitempty" yaml:"enablePrivilegedInitContainer,omitempty"`
 	LogLevel                      string `json:"logLevel,omitempty" yaml:"logLevel,omitempty"`
+	EnvoyImage                    string `json:"envoyImage,omitempty" yaml:"envoyImage,omitempty"`
```
Comment (snehachhabria) [inline pkg/apis/config/v1alpha1/mesh_config.go:28]: You will need to run `make codegen` to update the `zz_generated.deepcpoy.go` file as well

Comment (jaellio) [reply pkg/apis/config/v1alpha1/mesh_config.go:28]: EnvoyImage does not need to be added to `zz_generated.deepcopy.go` so `make codegen` will not update the file

**Pull request merged**